Add function editor to expression widget

NathanW2 opened the pull request:
**Note: For review and feedback. Not for merging yet**

This PR moves the custom function text area to a tab in the expression widget and allows for loading and saving of expression function files.  Expression function files are just normal Python files that live in `qgis2/functions` and are loaded at QGIS load time (not included in this PR yet).  Existing function files can be loaded from the combo box at the top of the widget.

![screenshot from 2015-01-12 18 43 27](https://cloud.githubusercontent.com/assets/381660/5700536/635afc02-9a8b-11e4-8053-19a20036abf0.png)

Pressing the run button will run the Python code in the current session (no syntax or error handling yet) but also save the current file.  Changing the text in the combo box will create a new file if not found or save over the existing one.  Changing the combo box value will load that file into the editor if found.

The rational for this change is because after I added the custom function text edit area it became apparent that you could not access the Python code you had just written (doh!).  I toyed with the idea of being able to store the Python code with the function and allow loading function code from the function tree however that lead to many user issues after testing like:
- Allows for more then one function per file. How do you load just the function you selected in the tree?
- Any smart file naming or auto saving/restoring felt really hacky
- Not a lot of room for more advanced logic 

Talking to Martin and Nyall this seemed like the smarter move.  It moves the focus more onto a code editor where you can define custom functions rather then a custom Python function (think ArcGIS) which the old UI implied 

This new method seems more clear to me as it allows for function grouping into single files, sharing of function files, larger code editor, clearer intent.

Feel free to try the code and leave any feedback. It's still not 100% yet but it does work.

NathanW2 committed:
Move custom function code editor to new tab

sgillies commented:
@NathanW2 From that example it looks like a "function file" is the same thing as a [Python module](https://docs.python.org/2/tutorial/modules.html)? That's a good thing. Is there more to a function file than that?

NathanW2 commented:
Yeah just a normal Python module. Not sure why I used that term.

On Tue, 13 Jan 2015 2:03 am Sean Gillies <notifications@github.com> wrote:

> @NathanW2 <https://github.com/NathanW2> From that example it looks like a
> "function file" is the same thing as a Python module
> <https://docs.python.org/2/tutorial/modules.html>? That's a good thing.
> Is there more to a function file than that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/1827#issuecomment-69592899>.
>

anitagraser commented:
So once the user adds a function in the function editor, it will then be available in the function tree in the expression tab?

NathanW2 commented:
Yes and it will be loaded every time qgis starts.

On Tue, 13 Jan 2015 7:04 am Anita Graser <notifications@github.com> wrote:

> So once the user adds a function in the function editor, it will then be
> available in the function tree in the expression tab?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/1827#issuecomment-69645574>.
>

NathanW2 commented:
@anitagraser here is a video https://www.youtube.com/watch?v=U2bopTw_BGg

NathanW2 committed:
Use Python console icons

NathanW2 committed:
New qgis.user module to handle loading of expressions and start up logic

Move saved functions into expressions folder in users home folder inside
python

Saved functions are loaded at loadtime

NathanW2 commented on python/user.py:
```diff
@@ -0,0 +1,36 @@
+    load_user_expressions([expressionspath])
+
+
+def load_user_expressions(paths):
```
@wonder-sk thoughts on this user file we talked about.  startup is called from C++ and then it handles the rest of the loading in Python.

wonder-sk replied:
Not sure about this one...

I would prefer not to add more paths to sys.path (I think it is a bad practice that pollutes the global namespace). Could we just have some generic approach where the user.py would make all subdirs (python packages) inside ~/.qgis2/python appear like they were located under qgis.user module? E.g. ~/.qgis2/python/expressions would be available as qgis.user.expressions

Then specific things like loading expression functions could be as simple as "from qgis.user.expressions import *" ?

Instead of "import startup" it seems that execfile() may be a better choice...

NathanW2 commented on src/python/qgspythonutilsimpl.cpp:
```diff
@@ -169,8 +178,7 @@ void QgsPythonUtilsImpl::initPython( QgisInterface* interface )
 
-  QString startuppath = homePythonPath() + " + \"/startup.py\"";
-  runString( "if os.path.exists(" + startuppath + "): from startup import *\n" );
+  runString( QString("qgis.user.startup(%1)").arg( homePythonPath() ) );
```
@wonder-sk new qgis.user module is called from here to kick off the Python user startup process.

wonder-sk replied:
I would put import qgis.user after initInterface() - to be the last thing done within initPython(). The call to qgis.user.startup(homedir) seems redundant as the code can be run when the module is first imported - and the homePythonPath() can be retrieved from QgsApplication...

NathanW2 committed:
Move logic to import of qgis.user. Add expressions package to qgis.user

NathanW2 committed:
exec startup.py. Create expressions folder and __init__ file

NathanW2 commented on src/gui/qgsexpressionbuilderwidget.cpp:
```diff
@@ -54,71 +56,18 @@ QgsExpressionBuilderWidget::QgsExpressionBuilderWidget( QWidget *parent )
+
+  updateFunctionTree();
+
+  mFunctionsPath = QgsApplication::qgisSettingsDirPath() + QDir::separator() + "python" + QDir::separator() + "expressions";
```
@wonder-sk  Do you think I should use `qgis.user` here to get the path?  Is that good for standalone applications.  I guess they can have a qgis.user module and set the path but that feels a bit hacky.

wonder-sk replied:
I would prefer ```qgis.user``` here - it should be fairly easy to set up even for standalone applications.

Unrelated idea - what about hiding the python editor tab when python is not enabled?

NathanW2 replied:
> Unrelated idea - what about hiding the python editor tab when python is
not enabled?

Yep thinking the same thing.

On Thu Jan 15 2015 at 1:18:21 PM Martin Dobias <notifications@github.com>
wrote:

> In src/gui/qgsexpressionbuilderwidget.cpp
> <https://github.com/qgis/QGIS/pull/1827#discussion-diff-22988552>:
>
> >
> >    txtExpressionString->setFoldingVisible( false );
> > -//  customFunctionBotton->setVisible( QgsPythonRunner::isValid() );
> > -  txtPython->setVisible( false );
> > -  cgbCustomFunction->setCollapsed( true );
> > -  txtPython->setText( "@qgsfunction(args=-1, group='Custom')\n"
> > -                      "def func(values, feature, parent):\n"
> > -                      "    return str(values)" );
> > +
> > +  updateFunctionTree();
> > +
> > +  mFunctionsPath = QgsApplication::qgisSettingsDirPath() + QDir::separator() + "python" + QDir::separator() + "expressions";
>
> I would prefer qgis.user here - it should be fairly easy to set up even
> for standalone applications.
>
> Unrelated idea - what about hiding the python editor tab when python is
> not enabled?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/1827/files#r22988552>.
>

NathanW2 committed:
Get expressionpath from qgis.user. Disable Python if not enabled

NathanW2 committed:
Fix expressions path loading. Add sip bindings and comments

NathanW2 committed:
Don't include __init__ file

NathanW2 commented:
@wonder-sk anything you want to see fixed before a merge?

wonder-sk commented:
I haven't tried it but the code looks good to me - feel free to merge

NathanW2 commented:
Merged in 59162bc
